what-is-chalk.mdx

```diff
@@ -18,215 +18,467 @@ import { PyDiffEditor } from '@/components/Editor'
+
+Chalk analyzes type annotations to build a directed acyclic graph (DAG) of feature dependencies.
+At inference time, query plans tailored to your input/output schema are generated dynamically by slicing this DAG into sub-graphs.
+This ensures that only the features needed are fetched--optimizing for speed, precision, and cost.
+
+Each node in the DAG is a Chalk expression or resolver (SQL / Python):
+
+- An expression e.g. `is_expensive_purchase: bool = _.amount > 100`
+- A SQL file that connects to an underlying data store like Athena
+- Python resolver i.e. ordinary Python functions (Pandas, httpx, or LLM calls)
+
+Here's an example of a generated query plan:
+
+![Chalk query plan](/img/docs-what_is_chalk-query_plan.png)
+
+### How does Chalk resolve features? What are Chalk resolvers?
+
+Chalk connects directly to your existing infrastructure and underlying data stores (Athena, Bigquery, Postgres, Iceberg catalog, etc.) with [SQL resolvers](https://docs.chalk.ai/docs/sql-resolvers) and to APIs (microservices, 3rd part clients, LLMs) with [Python resolvers.](https://docs.chalk.ai/docs/python-resolvers)
+Resolvers make it easy to integrate a wide variety of data sources, join them together, and use them in inference.
+
+Decoupling feature logic from ETL pipelines unlocks
+
+- faster iteration cycles
+- on-demand just-in-time inference
+- reproducible and testable features (with point-in-time accuracy)
 
 <IntegrationTypes />
 
----
+### How does Chalk orchestrate and manage features?
 
-## Execute
+Chalk has built-in support for feature engineering workflows — there's no need to manage Airflow or orchestrate complicated streaming flows.
+Once you've defined features and resolvers, Chalk orchestrates them into flexible pipelines (slicing the DAG) that make both training and model execution easy.
 
-Once you've defined your features and resolvers, Chalk orchestrates
-them into flexible pipelines that make training and executing models easy.
+The `get_username` function below explicitly defines input dependencies through `User.email` and specifies the output feature type as `User.username`:
 
-Chalk has built-in support for feature engineering workflows --
-there's no need to manage Airflow or orchestrate complicated streaming flows.
-You can execute resolver pipelines with declarative caching,
-ingest batch data on a schedule, and easily make slow sources
-available online for low-latency serving.
+```py
+@online
+def get_username(
+    # the input type annotation specifies the required feature
+    email: User.email,
+    # the resulting feature is saved as username unto User class
+) -> User.username:
+    username = email.split("@")[0]
+    if "gmail.com" in email:
+        username = username.split("+")[0].replace(".", "")
+
+    return username.lower()
+```
+
+Although written in Python, this function compiles into a static Velox expression through Chalk's Symbolic Python Interpreter and executes natively in C++.
```

Comment:
we first really introduce Velox 100 lines below, so maybe this instead talks about going to Chalk Expressions, and then we later talk about how chalk expressions are fast b/c velox?